src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java

```diff
@@ -217,23 +221,44 @@ protected Object invokeStep(StepDescriptor d, Object args) {
      * @param args The arguments passed to the step.
      */
     protected Object invokeStep(StepDescriptor d, String name, Object args) {
-        final NamedArgsAndClosure ps = parseArgs(args, d);
-
-        CpsThread thread = CpsThread.current();
-
-        FlowNode an;
 
         // TODO: generalize the notion of Step taking over the FlowNode creation.
         boolean hack = d instanceof ParallelStep.DescriptorImpl || d instanceof LoadStep.DescriptorImpl;
 
-        if (ps.body == null && !hack) {
-            an = new StepAtomNode(exec, d, thread.head.get());
+        FlowNode an;
+        CpsThread thread = CpsThread.current();
+        boolean hasBody = false;
+        if (!hack) {
+            if  (args != null) {
+                if (args instanceof NamedArgsAndClosure) {
+                    if (((NamedArgsAndClosure) args).body != null) {
+                        hasBody = true;
+                    }
+                } else if (args instanceof Object[]) {
+                    Object[] array = (Object[]) args;
+                    if (array.length > 0 && array[array.length - 1] instanceof CpsClosure) {
+                        hasBody = true;
+                    }
+                }
+            }
+        }
```

Comment:
Maybe make these a private method?  Otherwise add comments?

Reply:
added private method in 79d255a8427bc41a575bb6f808b355dfaadecb3e

Reply:
method was removed in cc5b3b0678b9607e6cfd197b56b9ed0c4fff8c18